### PR TITLE
8214455: Relocate CDS archived regions to the top of the G1 heap

### PR DESCRIPTION
**Pull request by iklam**
When the runtime heap and dump time heap have different sizes, it's possible that the archived heap regions are mapped to the middle of the runtime heap. Because the archived heap regions are pinned, this could reduce the size of the largest "humongous" array allocation. In the worst case, the maximum allocatable array length may be half of the optimal value.

The fix is to relocate the archived regions to the top of the  G1 heap (currently archived regions are supported only by G1 on 64-bit JVM).

Note that usually the top of the G1 heap is placed just below the 32GB boundary. As a result, the archived heap regions are at the same location between run time and dump time, so no relocation is necessary.

In mach5 testing, we occasionally run into this problem (see https://bugs.openjdk.java.net/browse/JDK-8239089), probably because we fail to reserve the heap below 32GB due to address space layout randomization (ASLR).

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8214455](https://bugs.openjdk.java.net/browse/JDK-8214455): Relocate CDS archived regions to the top of the G1 heap

### Reviewers
 * [Thomas Schatzl](https://openjdk.java.net/census#tschatzl) (@tschatzl - **Reviewer**) ⚠️ Review applies to def7db7bc68846531a934aa55b91e103819dde19
 * [Calvin Cheung](https://openjdk.java.net/census#ccheung) (@calvinccheung - **Reviewer**) ⚠️ Review applies to bedd676ddcd7dbbed6080f66120140c49feda224

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/3349/head:pull/3349` \
`$ git checkout pull/3349`

Update a local copy of the PR: \
`$ git checkout pull/3349` \
`$ git pull https://git.openjdk.java.net/jdk pull/3349/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 3349`

View PR using the GUI difftool: \
`$ git pr show -t 3349`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/3349.diff">https://git.openjdk.java.net/jdk/pull/3349.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/memory/filemap.cpp [hunk 1]
```diff
@@ -1837,6 +1842,9 @@ void FileMapInfo::map_heap_regions_impl() {
       log_info(cds)("the desired range " PTR_FORMAT " - "  PTR_FORMAT, p2i(range.start()), p2i(range.end()));
       log_info(cds)("is outside of the heap " PTR_FORMAT " - "  PTR_FORMAT, p2i(CompressedOops::begin()), p2i(CompressedOops::end()));
       _heap_pointers_need_patching = true;
+    } else if (header()->heap_end() != CompressedOops::end()) {
+      log_info(cds)("CDS heap data need to be relocated to the end of the runtime heap to reduce fragmentation");
```
Comment (tschatzl) [inline src/hotspot/share/memory/filemap.cpp:1846]: s/need/needs

File: test/hotspot/jtreg/runtime/cds/appcds/cacheObject/HeapFragmentationTest.java [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+ * questions.
+ *
+ */
+
+/*
+ * @test
+ * @summary Relocate CDS archived regions to the top of the G1 heap
+ * @bug 8214455
+ * @requires vm.cds.archived.java.heap
+ * @requires (sun.arch.data.model == "64"  & os.maxMemory > 4g)
```
Comment (tschatzl) [inline test/hotspot/jtreg/runtime/cds/appcds/cacheObject/HeapFragmentationTest.java:30]: additional space before the &

**Review by tschatzl — APPROVED** [review]
Lgtm.

File: src/hotspot/share/memory/filemap.cpp (hunk 1 above)
Comment (iklam) [reply src/hotspot/share/memory/filemap.cpp]: Fixed.

File: test/hotspot/jtreg/runtime/cds/appcds/cacheObject/HeapFragmentationTest.java (hunk 2 above)
Comment (iklam) [reply test/hotspot/jtreg/runtime/cds/appcds/cacheObject/HeapFragmentationTest.java]: Fixed. Thanks for the review.

File: test/hotspot/jtreg/runtime/cds/appcds/cacheObject/HeapFragmentationTest.java [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+                                              "-XX:+IgnoreUnrecognizedVMOptions",
+                                              "-XX:+G1ExitOnExpansionFailure");
+
+        // First, make sure the test runs without CDS
+        TestCommon.runWithoutCDS(TestCommon.concat(execArgs, runTimeHeapSize, mainClass, BUFF_SIZE))
+          .assertNormalExit(successOutput);
+
+        // Run with CDS. The archived heap regions should be relocated to avoid fragmentation.
+        TestCommon.run(TestCommon.concat(execArgs, runTimeHeapSize, mainClass,  BUFF_SIZE))
+          .assertNormalExit(successOutput);
```
Comment (calvinccheung) [inline test/hotspot/jtreg/runtime/cds/appcds/cacheObject/HeapFragmentationTest.java:84]: The assertNormalExit line should be indented four spaces relative to the previous line.

**Review by calvinccheung — APPROVED** [review]
Looks good. Just one minor nit in the test case.

**Comment by iklam** [comment]
Thanks @tschatzl and @calvinccheung for the review.
/integrate

**Pull request closed without merging**